cli/azd/pkg/ioc/context.go

```diff
@@ -0,0 +1,23 @@
+package ioc
```

Comment:
Why do we need to register the IoC container with the context and then pluck it out later?  It looked like there was only one component that consumed this (the middleware runner) and one component (the cobra builder) that set it, and the cobra builder calls the middleware runner.  It looked like there could be an indirect dependency via RunChildAction (which the build command calls) but it feels like both of these could have the IoC container or a service locator interface injected and not worry about using the context in this manner.

Would that work?

Reply:
The middleware runner is a singleton right now and stores the middleware registrations and some action results.  It does not currently have a direct coupling to a container.  When an action is run, it needs to resolve actions from different containers depending on the currnent executing scope.

In the case of running from the `cobra_builder` this is the new scope created for each `azd` command.  For child actions it is the current executing scope (but could also be a new scope if we want).

If we don't want to store the container in the go context, then we would need to open up the `RunAction()` function to accept a service locator as a parameter which is fine just didn't seem as clean.

I'll think on this a bit more and let you know where I land.

Reply:
I figured something out - take a peek.